PowerEditor/src/MISC/Common/Common.cpp

```diff
@@ -1158,38 +1158,84 @@ bool isCertificateValidated(const generic_string & fullFilePath, const generic_s
-			&& (buffer != NULL)                                 // check if buffer is not NULL
-			&& (wcsstr(buffer, TEXT("notepad++.exe")) == NULL); // check association with notepad++
-        
-	}
-	return isAssoCommandExisting;
-}
\ No newline at end of file
+
+	// Clean up.
+	if (pSignerInfo != NULL) LocalFree(pSignerInfo);
+	if (pCertContext != NULL) CertFreeCertificateContext(pCertContext);
+	if (hStore != NULL) CertCloseStore(hStore, 0);
+	if (hMsg != NULL) CryptMsgClose(hMsg);
+	if (szName != NULL) LocalFree(szName);
+
+	return isOK;
+}
+
+bool isAssoCommandExisting(LPCTSTR FullPathName)
+{
+	bool isAssoCommandExisting = false;
+
+	bool isFileExisting = PathFileExists(FullPathName) != FALSE;
+
+	if (isFileExisting)
+	{
+		PTSTR ext = PathFindExtension(FullPathName);
+
+		HRESULT hres;
+		wchar_t buffer[MAX_PATH] = TEXT("");
+		DWORD bufferLen = MAX_PATH;
+
+		// check if association exist
+		hres = AssocQueryString(ASSOCF_VERIFY|ASSOCF_INIT_IGNOREUNKNOWN, ASSOCSTR_COMMAND, ext, NULL, buffer, &bufferLen);
+        
+        isAssoCommandExisting = (hres == S_OK)                  // check if association exist and no error
+			&& (buffer != NULL)                                 // check if buffer is not NULL
+			&& (wcsstr(buffer, TEXT("notepad++.exe")) == NULL); // check association with notepad++
+        
+	}
+	return isAssoCommandExisting;
+}
+
+#ifndef _WIN64
+static ULONGLONG filetime_to_time_ull(const FILETIME* ft)
+{
+    ULARGE_INTEGER ull;
+    ull.LowPart = ft->dwLowDateTime;
+    ull.HighPart = ft->dwHighDateTime;
+    return (ull.QuadPart / 10000000ULL - 11644473600ULL);
+}
+
+int winapi_wstat(wchar_t const* _FileName, struct _stat* _Stat)
+{
+    // Starting from Visual Studio 2015, _wstat always returns -1 in Windows XP.
+    // So here is a WinAPI-based implementation of _wstat.
+    int nResult = -1;
+    DWORD dwAttr = ::GetFileAttributes(_FileName);
+    if (dwAttr != INVALID_FILE_ATTRIBUTES)
+    {
+        HANDLE hFile = ::CreateFile(_FileName, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
+        if (hFile != INVALID_HANDLE_VALUE)
+        {
+            LARGE_INTEGER fileSize;
+            FILETIME creationTime, accessTime, writeTime;
+            if (::GetFileSizeEx(hFile, &fileSize) &&
+                ::GetFileTime(hFile, &creationTime, &accessTime, &writeTime))
+            {
+                ::ZeroMemory(_Stat, sizeof(struct _stat));
+                _Stat->st_atime = static_cast<decltype(_Stat->st_atime)>(filetime_to_time_ull(&accessTime));
+                _Stat->st_ctime = static_cast<decltype(_Stat->st_ctime)>(filetime_to_time_ull(&creationTime));
+                _Stat->st_mtime = static_cast<decltype(_Stat->st_mtime)>(filetime_to_time_ull(&writeTime));
+                _Stat->st_size = static_cast<decltype(_Stat->st_size)>(fileSize.QuadPart);
+                _Stat->st_mode = _S_IREAD | _S_IEXEC;

```

Comment:
Are directories executable?

Reply:
As https://www.gnu.org/software/libc/manual/html_node/Permission-Bits.html says:
S_IEXEC : Execute (for ordinary files) or search (for directories) permission bit for the owner of the file